client/my-sites/site-settings/site-icon-setting/index.jsx

```diff
@@ -40,11 +51,45 @@ class SiteIconSetting extends Component {
 
 	showModal = () => this.toggleModal( true );
 
-	setSiteIcon = ( media ) => {
-		// [TODO]: Handle setting site icon
-		console.log( media ); // eslint-disable-line no-console
+	editSelectedMedia = ( media ) => {
+		if ( media ) {
+			this.props.onEditSelectedMedia();
+		}
+	};
 
+	setSiteIcon = ( error, blob ) => {
 		this.hideModal();
+		this.props.resetAllImageEditorState();
+
+		if ( error || ! blob ) {
+			return;
+		}
+
+		const { siteId } = this.props;
+		const selectedItem = head( MediaLibrarySelectedStore.getAll( siteId ) );
+		if ( ! selectedItem ) {
+			return;
+		}
+
+		const transientMediaId = uniqueId( 'site-icon' );
+
+		MediaActions.add( siteId, {
+			ID: transientMediaId,
+			fileName: `cropped-${ selectedItem.file }`,
+			fileContents: blob
+		} );
+
+		const checkUploadComplete = () => {
+			const transientMedia = MediaStore.get( siteId, transientMediaId );
+			if ( isItemBeingUploaded( transientMedia ) ) {
+				return;
+			}
+
+			MediaStore.off( 'change', checkUploadComplete );
+			alert( transientMedia.ID );
+		};
+
+		MediaStore.on( 'change', checkUploadComplete );
```

Comment:
Toying with this a bit, I wonder if this becomes a common pattern that we create an abstraction for it, one which perhaps could accommodate both a Redux store subscription or an EventEmitter object (e.g. Flux store). Something like:

```js
untilStoreSatisfies( MediaStore, () => (
	! isItemBeingUploaded( MediaStore.get( siteId, transientMediaId ) )
) ).then( () => {
	alert( MediaStore.get( siteId, transientMediaId ).ID );
} );
```

```js
const { store } = this.context;
untilStoreSatisfies( store, () => (
	! isItemBeingUploaded( store.getState(), siteId, transientMediaId ) )
) ).then( () => {
	alert( getMediaItem( store.getState(), siteId, transientMediaId ).ID );
} );
```